prow/metrics/prowjobs/collector.go

```diff
@@ -0,0 +1,172 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prowjobs
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/tools/cache"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+type prowJobLifecycleCollector struct {
+	HistogramVec *prometheus.HistogramVec
+}
+
+func (pjc *prowJobLifecycleCollector) Update(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) {
+	if oldJob == nil {
+		return
+	}
+	if oldJob.Status.State != newJob.Status.State {
+
+		labels := getJobLabel(oldJob, newJob)
+		histogram, err := pjc.HistogramVec.GetMetricWithLabelValues(labels.values()...)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to get a histogram for a prowjob")
+			return
+		}
+
+		var oldTime metav1.Time
+		var newTime metav1.Time
+
+		switch oldJob.Status.State {
+		case prowapi.TriggeredState:
+			oldTime = oldJob.CreationTimestamp
+		case prowapi.PendingState:
+			if oldJob.Status.PendingTime == nil {
+				return
+			}
+			oldTime = *oldJob.Status.PendingTime
+		default:
+			return
+		}
+
+		switch newJob.Status.State {
+
+		case prowapi.FailureState, prowapi.SuccessState, prowapi.ErrorState, prowapi.AbortedState:
+			if newJob.Status.CompletionTime == nil {
+				return
+			}
+			newTime = *newJob.Status.CompletionTime
+		case prowapi.PendingState:
+			if newJob.Status.PendingTime == nil {
+				return
+			}
+			newTime = *newJob.Status.PendingTime
+		default:
+			return
+		}
+
+		histogram.Observe(newTime.Sub(oldTime.Time).Seconds())
+	}
+}
+
+// NewProwJobLifecycleCollector creates histograms which can track the timespan between ProwJob state transitions.
+// The histograms are based on the job name, the old job state and the new job state.
+// Data is collected by hooking itself into the prowjob informer.
+// The collector will never record the same state transition twice, even if reboots happen.
+func NewProwJobLifecycleCollector(informer cache.SharedIndexInformer) *prowJobLifecycleCollector {
+	lifecycleCollector := &prowJobLifecycleCollector{
+		HistogramVec: newHistogramVec(),
+	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldJob, newJob interface{}) {
+			lifecycleCollector.Update(oldJob.(*prowapi.ProwJob), newJob.(*prowapi.ProwJob))
+		},
+	})
+	return lifecycleCollector
+}
+
+func getJobLabel(oldJob *prowapi.ProwJob, newJob *prowapi.ProwJob) jobLabel {
+	jl := jobLabel{jobNamespace: newJob.Namespace, jobName: newJob.Spec.Job, jobType: string(newJob.Spec.Type), state: string(newJob.Status.State), last_state: string(oldJob.Status.State)}
```

Comment:
nit: can you put these on separate lines, so it's not one single long one

Reply:
Done